### PR TITLE
feat: Support api version 2 for nuget datasource

### PR DESCRIPTION
**Pull request by Elufimov**
This PR adds support for api version 2 for nuget. This will allow to use proget, witch we currently use,  with renovate. I set defaults to nuget.org therefore there are no breaking changes. But this is my first interaction with js project and I will be happy to fix any comments.

### CODE REVIEW

File: lib/config/definitions.js [hunk 1]
```diff
@@ -1397,7 +1397,13 @@ const options = [
     stage: 'package',
     type: 'json',
     default: {
-      fileMatch: ['\\.csproj$'],
+      fileMatch: ['(^|/)*\\.csproj$'],
```
Comment (rarkins) [inline lib/config/definitions.js]: What are you trying to achieve with this change?

File: lib/config/definitions.js [hunk 2]
```diff
@@ -1397,7 +1397,13 @@ const options = [
...
+      nugetFeeds: [
```
Comment (rarkins) [inline lib/config/definitions.js]: I think `registryUrls` is an existing concept we have that might fit this. Would these ever be "autodetected" from the repository, or users need to configure it manually in Renovate config?

**Comment by rarkins** [comment]
@Elufimov thanks for the PR. Could servers or api versions ever be detected from within the repo, or do you expect they'd always need to be configured manually in Renovate config?

Also, would it be expected that you'd ever have a *mix* of v2 and v3 feeds and try both?

File: lib/config/definitions.js (hunk 1 above)
Comment (Elufimov) [reply lib/config/definitions.js]: Nothing, this is not a part of request. I will revert it.

**Comment by rarkins** [comment]
@Elufimov First I need to double check if `registryUrls` can satisfy requirements

File: lib/config/definitions.js (hunk 2 above)
Comment (rarkins) [reply lib/config/definitions.js]: OK, here's what I'd like to do:

1. Do not use new `nugetFeeds` option (remove it)
2. Use existing `registryUrls` array instead
3. Do not explicitly define the default value inside `nuget` config. We can simply default to it inside datasource/nuget instead
4. Default to `v3`. But if you want to define a v2 feed then do it like `https://https://version2#v2`. The nuget datasource will look for this URL fragment and strip it before using it. Using `v3` should be optional.

**Review by rarkins — CHANGES_REQUESTED** [review]
Use `registryUrls` instead of `nugetFeeds`

File: lib/config/definitions.js (hunk 1 above)
Comment (Elufimov) [reply lib/config/definitions.js]: Reverted

File: lib/config/definitions.js (hunk 2 above)
Comment (Elufimov) [reply lib/config/definitions.js]: 1-3 done and in 4 I made the auto detection system.

File: lib/datasource/nuget/index.js [hunk 3]
```diff
@@ -1,51 +1,190 @@
-const parse = require('github-url-from-git');
```
Comment (rarkins) [inline lib/datasource/nuget/index.js]: Shouldn't we still use this to parse the URLs?

**Comment by rarkins** [comment]
Thanks @Elufimov. Can you describe how much testing on "real" repositories you've been able to do? Mostly I want to make sure there's no regression failures for the existing users of the default registry.

File: lib/datasource/nuget/index.js (hunk 3 above)
Comment (Elufimov) [reply lib/datasource/nuget/index.js]: I returned `sourceUrl` verification if https://github.com/renovatebot/renovate/pull/3148/commits/dbb98ecf5c979c41fb4464436ac07dc3170bc6de

**Comment by Elufimov** [comment]
I tested against one of our application. We use private gitlab instance. The project contains both public dependencies from nuget.org and private dependencies from our local proget instance. I used this config:
```json
{
  "extends": [
    "config:base",
    ":rebaseStalePrs",
    "group:linters",
    ":disableRateLimiting"
  ],
  "semanticCommitScope": "deps",
  "docker": {
    "enabled": false
  },
  "prBodyColumns": ["Package", "Update", "Type", "Change", "References"],
  "token": "some_token",
  "platform": "gitlab",
  "repositories": ["Group/Repository"],
  "endpoint": "https://gitlab.local/api/v4",
  "log-level": "debug",
  "nuget": {
    "registryUrls": [
      "https://api.nuget.org/v3/index.json",
      "http://proge.local/nuget/feed_name"
    ]
  },
  "recreateClosed": true,
  "packageRules": [
    {
      "packagePatterns": ["^PravoRu"],
      "groupName": "PravoRu packages",
      "automerge": false
    }
  ]
}
``` 
As result I had multiple pull requests. It seems everything working as expected. However I can do other tests if you want.

**Comment by rarkins** [comment]
@Elufimov I was able to test this PR out on an example Nuget repo just now. The good news is that it finds and creates the same number of PRs as before. The only problem I found is that the `sourceUrl` is missing and hence the possibility to retrieve release notes. Here's an example PR that's missing release notes in your PR but has them present in `master`: https://github.com/renovate-tests/dotnet-example/pull/2

This may be because of the change of "endpoints" used to retrieve metadata. Can you look into it and tell me if you think there's any way to still get the sourceUrl? Maybe for example we query the "old" endpoint just to get the metadata and then we cache it for a long period to optimize it.

**Comment by Elufimov** [comment]
@rarkins Oh, I see the problem. I broke `sourceUrl` by using `sourceUrl` from package description instead of `repository@url` from nuspec. I will fix it.

**Comment by Elufimov** [comment]
> Maybe for example we query the "old" endpoint just to get the metadata and then we cache it for a long period to optimize it.
The issue with "old" endpoint `v3-flatcontainer` is that it not universal. Myget do not have it for example. Is it a good idea to download pkg, unzip it and read nuspec from it? I will look for more options tomorrow.

**Comment by Elufimov** [comment]
I also could add back old algorithm only for nuget.org. In that case there will not be any breaking changes for current users.

**Comment by Elufimov** [comment]
I have done more research and I do not see `sourceUrl` in v2 or in v3 api. I see two options:
1) Download pkg, extract nuspec from it and then get `sourceUrl`
2) Check if feed is a `https://api.nuget.org/v3/index.json` and then use `https://api.nuget.org/v3-flatcontainer/${lookupName.toLowerCase()}/${res.versions.pop()}/${lookupName.toLowerCase()}.nuspec` else try to use `ProjectUrl`
Second option is less time consuming for implementing.

**Comment by rarkins** [comment]
@Elufimov I like the second option too.

**Comment by Elufimov** [comment]
@rarkins I implemented a second option as we discussed.

**Comment by fgreinacher** [comment]
I'd love to have this feature, thanks for the work! I'm wondering why the NuGet feeds need to be configured again in the renovate config, when they will usually already be stored in a `NuGet.config` within the repository. Furthermore, there needs to be way to configure credentials for accessing private feeds. Would you consider adding at least basic support for this by reading additional HTTP headers e.g. from an environment variable?

**Comment by Elufimov** [comment]
NuGet.config is not a mandatory part of project repository. It could be there. For example we do not store it in ours repositories. As for authentication I would not want to include this in this pr. It could be done in separate request.

**Comment by fgreinacher** [comment]
OK, no problem :) I'm wondering whether authentication could be supplied via the existing [host rules](https://renovatebot.com/docs/configuration-options/#hostrules) configuration?!

**Comment by Elufimov** [comment]
If I found the right example https://github.com/renovatebot/renovate/blob/master/lib/datasource/docker/index.js#L54 it could be relatively easy adopted.

**Comment by rarkins** [comment]
I've raised a couple of feature requests for (a) feed detection, and (b) authentication, you can see them linked now to this PR.

**Pull request merged**